modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/latch/ExchangeLatchManager.java

```diff
@@ -99,6 +101,13 @@
     @GridToStringExclude
     private final ConcurrentMap<AffinityTopologyVersion, ClusterNode> joinedNodes = new ConcurrentHashMap<>();
 
+    /** This map holds version on which ack message have been received before local latch was created. */
+    private final ConcurrentSkipListMap<AffinityTopologyVersion, GridFutureAdapter<Object>> awaitedLatchVersions =
```

Comment:
ExchangeLatchManager can handle several latches for the same topologyVersion, see CompletableLatchUid.
We need to handle correctly all of them.